variable_set.go

```diff
@@ -0,0 +1,255 @@
+type VariableSets interface {
+	// List all the variable sets within an organization.
+	List(ctx context.Context, organization string, options VariableSetListOptions) (*VariableSetList, error)
+
+	// Create is used to create a new variable set.
+	Create(ctx context.Context, organization string, options VariableSetCreateOptions) (*VariableSet, error)
+
+	// Read a variable set by its ID.
+	Read(ctx context.Context, variableSetID string) (*VariableSet, error)
+
+	// Update an existing variable set.
+	Update(ctx context.Context, variableSetID string, options VariableSetUpdateOptions) (*VariableSet, error)
+
+	// Delete a variable set by ID.
+	Delete(ctx context.Context, variableSetID string) error
+
+	// Assign a variable set to workspaces
+	Assign(ctx context.Context, variableSetID string, options VariableSetAssignOptions) (*VariableSet, error)
+}
+
+type variableSets struct {
+	client *Client
+}
+
+type VariableSetList struct {
+	*Pagination
+	Items []*VariableSet
+}
+
+type VariableSet struct {
+	ID          string `jsonapi:"primary,varsets"`
+	Name        string `jsonapi:"attr,name"`
+	Description string `jsonapi:"attr,description"`
+	Global      bool   `jsonapi:"attr,global"`
+
+	// Relations
+	Workspaces []*Workspace           `jsonapi:"relation,workspaces,omitempty"`
+	Variables  []*VariableSetVariable `jsonapi:"relation,vars,omitempty"`
+}
+
+type VariableSetListOptions struct {
+	ListOptions
+}
+
+func (o VariableSetListOptions) valid() error {
+	return nil
+}
+
+// List all Variable Sets in the organization
+func (s *variableSets) List(ctx context.Context, organization string, options VariableSetListOptions) (*VariableSetList, error) {
+	if !validStringID(&organization) {
+		return nil, ErrInvalidOrg
+	}
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	u := fmt.Sprintf("organizations/%s/varsets", url.QueryEscape(organization))
+	req, err := s.client.newRequest("GET", u, &options)
+	if err != nil {
+		return nil, err
+	}
+
+	vl := &VariableSetList{}
+	err = s.client.do(ctx, req, vl)
+	if err != nil {
+		return nil, err
+	}
+
+	return vl, nil
+}
+
+// VariableSetCreateOptions represents the options for creating a new variable set within in a organization.
+type VariableSetCreateOptions struct {
+	// Type is a public field utilized by JSON:API to
+	// set the resource type via the field tag.
+	// It is not a user-defined value and does not need to be set.
+	// https://jsonapi.org/format/#crud-creating
+	Type string `jsonapi:"primary,vars"`
+
+	// The name of the variable set.
+	// Affects variable precedence when there are conflicts between Variable Sets
+	// https://www.terraform.io/cloud-docs/api-docs/variable-sets#apply-variable-set-to-workspaces
+	Name *string `jsonapi:"attr,name"`
+
+	// A description to provide context for the variable set.
+	Description *string `jsonapi:"attr,description"`
+
+	// If true the variable set is considered in all runs in the organization.
+	Global *bool `jsonapi:"attr,global"`
```

Comment:
Since these two are not required by the API, I think they should be annotated omitempty so that null isn't passed to the API. I also noticed that you required Global in line 113 so that null wouldn't be passed but the API doesn't have the same requirement and defaults to false. If this were omitempty, I don't think you'd have to make that nil check

Reply:
Good call!